allel/io/vcfreader.py

```diff
@@ -0,0 +1,255 @@
+        check out https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3042176/
+        and https://samtools.github.io/hts-specs/tabix.pdf
+        for more information.
+
+        Parameters
+        ----------
+        begin: int
+            chromosome position begin
+        end: int
+            chromosome position end
+        n_lvls: int, optional
+            cluster level, for tabix, set to 5
+        min_shift: int, optional
+            minimum shift, for tabix, set to 14
+
+        Returns
+        -------
+        generator
+
+        """
+        begin, end = begin, end
+        t, s = 0, min_shift + (n_lvls << 1) + n_lvls
+        for l in range(n_lvls + 1):
+            b, e = t + (begin >> s), t + (end >> s)
+            n = e - b + 1
+            for k in range(b, e + 1):
+                yield k
+                n += 1
+            t += 1 << ((l << 1) + l)
+            s -= 3
+
+    def _parse_index(self, index):
+        """
+        read and parse the index file, set the virtual offset for the target region,
+        check out https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3042176/
+        and https://samtools.github.io/hts-specs/tabix.pdf
+        for more information.
+
+        Parameters
+        ----------
+        index: str
+            path for the index file
+
+        """
+        with gzip.open(index, 'rb') as f:
+            magic = f.read(4)
+            if magic != b'TBI\x01':  # check magic
+                raise RuntimeError('invalid tabix index file: {}'.format(index))
+            header = np.frombuffer(f.read(4 * 8), dtype=np.int32)
+            if header[1] != 2:  # check format
+                raise RuntimeError('invalid VCF index file: {}'.format(index))
+            # this is the comment marker use to determine VCF header, usually set to b'#'
+            self._meta = chr(header[5]).encode('ascii')
+            try:
+                rid = f.read(header[7]).split(b'\x00').index(self._chrom)
+            except ValueError:
+                # contig name is not exist, read header only
+                self._chunk_begin = -1
+                return
+            # seek to our target index
+            for _ in range(rid):
+                for _ in range(struct.unpack('<i', f.read(4))[0]):  # n_bins
+                    f.seek(4, os.SEEK_CUR)  # bin_key
+                    n_chunk = struct.unpack('<i', f.read(4))[0]  # n_chunk
+                    f.seek(8 * 2 * n_chunk, os.SEEK_CUR)  # cnk_beg, cnk_end
+                n_intv = struct.unpack('<i', f.read(4))[0]  # n_intv
+                f.seek(8 * n_intv, os.SEEK_CUR)  # ioff
+            # read our target index
+            bidx = {bin_key: None for bin_key in self.reg2bins(self._begin, self._end)}
+            for _ in range(struct.unpack('<i', f.read(4))[0]):  # n_bins
+                bin_key = struct.unpack('<I', f.read(4))[0]  # bin
+                n_chunk = struct.unpack('<i', f.read(4))[0]  # n_chunk
+                if bin_key in bidx:
```

Comment:
``n_chunk`` might be zero? If so, could use that to simplify a little...

```suggestion
                if bin_key in bidx and n_chunk > 0:
```

Reply:
I am not sure would `n_chunk` be zero? But `for` can use as `if`, and a judge do nothing.